src/ServiceControl.Plugin.Nsb5.Heartbeat/Heartbeats.cs

```diff
@@ -12,96 +12,159 @@
 
-        void SendStartupMessageToBackend(HostInformation hostInfo)
-        {
-            backend.Send(
-                new RegisterEndpointStartup
+            void SendEndpointStartupMessage(HostInformation hostInfo, DateTime startupTime, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    backend.Send(
+                        new RegisterEndpointStartup
+                        {
+                            HostId = hostInfo.HostId,
+                            Host = hostInfo.DisplayName,
+                            Endpoint = endpointName,
+                            HostDisplayName = hostInfo.DisplayName,
+                            HostProperties = hostInfo.Properties,
+                            StartedAt = startupTime
+                        }, ttlTimeSpan);
+                }
+                catch (Exception ex)
                 {
-                    HostId = hostInfo.HostId,
+                    Logger.Warn(string.Format("Unable to register endpoint startup with ServiceControl. Going to reattempt registration after {0}.", registrationRetryInterval), ex);
+
+                    Task.Delay(registrationRetryInterval, cancellationToken)
+                        .ContinueWith(t => SendEndpointStartupMessage(hostInfo, startupTime, cancellationToken), cancellationToken);
+                }
+            }
+
+            void SendHeartbeatMessage(HostInformation hostInfo)
+            {
+                var heartBeat = new EndpointHeartbeat
+                {
+                    ExecutedAt = DateTime.UtcNow,
+                    EndpointName = endpointName,
                     Host = hostInfo.DisplayName,
-                    Endpoint = Configure.Settings.EndpointName(),
-                    HostDisplayName = hostInfo.DisplayName,
-                    HostProperties = hostInfo.Properties,
-                    StartedAt = DateTime.UtcNow
-                }, ttlTimeSpan);
-        }
+                    HostId = hostInfo.HostId
+                };
 
-        void ExecuteHeartbeat(HostInformation hostInfo)
-        {
-            var heartBeat = new EndpointHeartbeat
+                try
+                {
+                    backend.Send(heartBeat, ttlTimeSpan);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Warn("Unable to send heartbeat to ServiceControl:", ex);
+                }
+            }
+
+            public void Dispose()
             {
-                ExecutedAt = DateTime.UtcNow,
-                EndpointName = Configure.Settings.EndpointName(),
-                Host = hostInfo.DisplayName,
-                HostId = hostInfo.HostId
-            };
-            backend.Send(heartBeat, ttlTimeSpan);
-        }
+                try
+                {
+                    if (cancellationTokenSource != null)
+                    {
+                        cancellationTokenSource.Dispose();
```

Comment:
@SzymonPobiega I did it for the "important" note on MSDN daniel posted. I didn't dig deeper into that.